fix(add-caching): explicitly set targetDefaults for all scripts

fahslaj opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
- Explicitly sets `targetDefaults` for all scripts, not just those selected in the prompts.
- Updates how cacheable targets are configured to align with Nx 17 expectations.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This allows `lerna run` to correctly identify when the user has configured Nx via `lerna add-caching`, even if they didn't select any scripts to run in order. This also fixes `add-caching` for Nx 17.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This has been tested manually on a fresh workspace.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (change that has absolutely no effect on users)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

fahslaj commented on packages/lerna/src/commands/add-caching/index.ts:
```diff
@@ -55,16 +64,23 @@ class AddCachingCommand extends Command {
         message:
           "Which scripts need to be run in order? (e.g. before building a project, dependent projects must be built.)\n",
         choices: this.uniqueScriptNames,
+        default: existingTargetDefaults,
```
This allows inquirer to pre-select all targets that already have dependsOn configured

<img width="785" alt="image" src="https://github.com/lerna/lerna/assets/6913035/59afde67-9522-4a4d-9ba3-c9419aa85273">

fahslaj commented on packages/lerna/src/commands/add-caching/index.ts:
```diff
@@ -74,13 +90,13 @@ class AddCachingCommand extends Command {
         message:
           "Which scripts are cacheable? (Produce the same output given the same input, e.g. build, test and lint usually are, serve and start are not.)\n",
         choices: this.uniqueScriptNames,
+        default: existingCacheableOperations,
```
Same as above; this allows inquirer to pre-select all targets that have `"cache": true` already configured

fahslaj commented on packages/lerna/src/commands/add-caching/index.ts:
```diff
@@ -108,29 +126,7 @@ class AddCachingCommand extends Command {
-      // eslint-disable-next-line no-empty
-    } catch {}
-
-    nxJson.tasksRunnerOptions = nxJson.tasksRunnerOptions || {};
-    nxJson.tasksRunnerOptions["default"] = nxJson.tasksRunnerOptions["default"] || {};
-    nxJson.tasksRunnerOptions["default"].runner =
-      nxJson.tasksRunnerOptions["default"].runner || "nx/tasks-runners/default";
-    nxJson.tasksRunnerOptions["default"].options = nxJson.tasksRunnerOptions["default"].options || {};
-
-    if (nxJson.tasksRunnerOptions["default"].options.cacheableOperations) {
-      this.logger.warn(
-        "add-caching",
-        "The `tasksRunnerOptions.default.cacheableOperations` property already exists in `nx.json` and will be overwritten by your answers"
-      );
-    }
```
Adding the default task runner seems like excess config. Nx should default this for us when this is missing.

The meaningful removal here is of `cacheableOperations`, which moved to `targetDefaults`.

JamesHenry commented on website/docs/api-reference/configuration.md:
```diff
@@ -69,10 +69,7 @@ Find the available options in [the API docs](/docs/api-reference/commands).
-      "options": {
-        "cacheableOperations": ["build", "test"]
-      }
+      "runner": "nx/tasks-runners/default"
```
Shouldn’t we remove this here too for simplicity?

JamesHenry commented on website/docs/api-reference/configuration.md:
```diff
@@ -99,16 +98,16 @@ Find the available options in [the API docs](/docs/api-reference/commands).
 Everything in Nx is customizable, including running npm scripts. Most of the time you will either use the default runner
 or the `@nrwl/nx-cloud` runner.
 
-### cacheableOperations
```
We can remove the paragraph on runner above it’s outdated

JamesHenry commented on website/docs/concepts/how-caching-works.md:
```diff
@@ -199,7 +199,6 @@ option for the task runner in `nx.json`:
   "tasksRunnerOptions": {
     "default": {
       "options": {
-        "cacheableOperations": ["build", "test"],
```
This whole config is outdated too, cache directory is set at the top level, no needs for the object

fahslaj replied:
I've removed this whole section. We're pretty clear that it's Nx doing the caching in the docs and users can find the Nx docs if they need to do advanced things like changing the cache location.

I think we should minimize the amount Nx-specific docs on the Lerna site as they will get out of date quickly.

JamesHenry commented on website/docs/concepts/task-pipeline-configuration.md:
```diff
@@ -32,7 +32,6 @@ Note, you can also change the default in `nx.json`, like this:
     "default": {
       "runner": "nx/tasks-runners/default",
       "options": {
-        "cacheableOperations": [],
```
Same again, outdated

fahslaj replied:
Removed

JamesHenry commented on packages/lerna/src/commands/add-caching/index.ts:
```diff
@@ -156,6 +159,22 @@ class AddCachingCommand extends Command {
+      // .nx/cache is already ignored - no need to update .gitignore
+    } catch (e) {
+      try {
+        await writeFile(joinPathFragments(workspaceRoot, ".gitignore"), "\n.nx/cache\n", { flag: "a+" });
```
I’m not familiar with passing this a+ flag, I would have gone for appendFile, I’m on mobile and very late but assuming this makes this equivalent?

fahslaj replied:
Yes, this is equivalent to using `appendFile`. I went ahead and changed it to use `appendFile` since that's more readable.

JamesHenry reviewed (changes requested):
See comments

fahslaj commented:
@JamesHenry I've updated the docs you've pointed out. I removed a couple of the outdated sections instead of just updating their config. To reduce the maintenance burden, I think we should keep the more advanced Nx configuration options to the Nx docs and link to the Nx site instead of duplicating them in the Lerna docs.
